### DISCUSSION TITLE
Snowflake Support

### DISCUSSION THREAD

**Discussion by datastx** [moj-analytical-services/splink#2303] [General]
@RobinL  Im looking at taking a swing at using this vs building something myself and my backend uses snowflake. I was wondering without digging in too far whats the level of effort to use snowflake as an engine since thats what I use and I imagine a lot of main stream private sector users use as well.

**Comment by datastx** [comment]
It looks like I just need to update the pyproject.toml, create a snowflake.py in the backends, add a __init__.py, database_api.py, dataframe_api.py in the backends folder and add a workflow to test snowflake? All this on its own fork I assume and then get some of the examples running using snowflake?

**Comment by ADBond** [comment]
Hello! We are probably not in a position to be able to support Snowflake _directly_, as we do not have access to it and have limited capacity. However, Splink should (hopefully) be set up to allow you to write a Snowflake extension that is relatively light-touch. You shouldn’t even need to fork Splink - things should work with an installed version of Splink + a fresh project, which should make the dev setup a bit more straightforward, and wouldn't need to directly keep in-sync with updates to Splink code in the main.

In terms of what is needed to get things working, you will need a custom `DatabaseAPI` (for managing interactions with the db), `SplinkDataFrame` (a couple of things for interacting with tables) and a `SplinkDialect` (which handles any dialect-specific SQL). So something like this:
```python
from splink.internals.database_api import DatabaseAPI
from splink.internals.dialects import SplinkDialect
from splink.internals.splink_dataframe import SplinkDataFrame

class SnowflakeAPI(DatabaseAPI):
    # overwrite required methods
    # most of the work will be here
    ...

class SnowflakeDataDrame(SplinkDataframe):
    # overwrite required methods
    ...

class SnowflakeDialect(SplinkDialect):
    # overwrite required methods
    ...
```
Probably the most useful blueprint will be [postgres](https://github.com/moj-analytical-services/splink/tree/master/splink/internals/postgres), as it doesn’t need many special cases, and it utilises sqlalchemy, so a lot of code would be similar if you went down that route also. If you can get the [getting started script](https://moj-analytical-services.github.io/splink/getting_started.html) working (possibly needing to change some of the comparisons depending on the fuzzy-string-matching capabilities of Snowflake) you should probably be good to go with most things Splink can offer.

If you do run into any difficulties, or if it seems that there may actually be some changes need to Splink to facilitate this extension then do please reach out and I’ll try to help as best I can.

**Reply by datastx** [reply]
Awesome and appreciate the insight. What Im hearing is maybe I create a repo that has splink as a dep as an install and write the corresponding abstractions mentioned above and I can make a lose link giving kudos etc where myself as a snowflake user will do what I can to support it and you're willing to give your 2 cents on some PRs?

**Reply by ADBond** [reply]
Yep, I think that sounds like a reasonable plan 👍

**Reply by bagerman** [reply]
@datastx I was wondering if you had a chance to follow on what @ADBond suggested and would not mind sharing it?

**Comment by ADBond** [comment]
FYI @bagerman (and @datastx) - I don't know what the progress is with adding Snowflake support (if any), but I have developed a [separate package for adding Clickhouse support to Splink](https://github.com/ADBond/splinkclickhouse), partly as something of a template for anyone wishing to add any other backends (and to confirm that adding support for backends can be done downstream without needing to make changes to Splink).

Some stuff is Clickhouse-specific, but it should broadly give an idea of the kinds of things you may need to do to get Splink working with Snowflake (or another backend).

**Comment by supermacro** [comment]
As someone new to Splink (and data matching more broadly), could I get some guidance on how to think about the following two options:

- using a snowflake backend db / api (our team is considering implementing one)
- running splink in spark (or in the case of snowflake; snowpark) environment

Are these two things mutually exclusive? e.g. if running in a spark-like environment, would I simply load partitions of the data on various nodes into duckdb for splink to then perform analysis? Or would I still need a snowflake backend if running inside of snowflake?

**Reply by RobinL** [reply]
In Splink, the main computations are delegated to the SQL backend - see [backends](https://moj-analytical-services.github.io/splink/topic_guides/splink_fundamentals/backends/backends.html?h=bac)

If you have Spark already, you probably want to use the Spark backend that handles linkage natively in Spark dataframes, without needing to convert to duckdb.

Alternatively you can try write a native snowflake backend, similar to Andy;'s clickhouse one [here](https://github.com/ADBond/splinkclickhouse).  But we don't have access to snowflake, so it'd be up to you to create it and make sure it works
